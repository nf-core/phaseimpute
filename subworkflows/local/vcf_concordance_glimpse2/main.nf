include { GLIMPSE2_CONCORDANCE        } from '../../../modules/nf-core/glimpse2/concordance'
include { GAWK                        } from '../../../modules/nf-core/gawk'
include { ADD_COLUMNS                 } from '../../../modules/local/add_columns'
include { GUNZIP                      } from '../../../modules/nf-core/gunzip'

workflow VCF_CONCORDANCE_GLIMPSE2 {

    take:
        ch_vcf_emul   // VCF file with imputed genotypes [ [id, panel, tool], vcf, csi]
        ch_vcf_truth  // VCF file with truth genotypes   [ [id, panel], vcf, csi]
        ch_vcf_freq   // VCF file with panel frequencies [ [panel], vcf, csi]
        ch_region     // Regions to process              [ [chr, region], region]

    main:

    ch_versions      = Channel.empty()
    ch_multiqc_files = Channel.empty()

    ch_concordance = ch_vcf_emul
        .map{metaIPTC, vcf, csi -> [metaIPTC.subMap("id"), metaIPTC, vcf, csi]}
        .combine(ch_vcf_truth
            .map{metaIPC, vcf, csi -> [ metaIPC.subMap("id"), vcf, csi ]}
            , by: 0
        )
        .combine(ch_vcf_freq)
        .combine(ch_region.map{[it[1]]}.collect().toList())
        .map{metaI, metaIPTC, emul, e_csi, truth, t_csi, metaP, freq, f_csi, regions ->
            [metaIPTC, emul, e_csi, truth, t_csi, freq, f_csi, [], regions]
        }

    GLIMPSE2_CONCORDANCE (
        ch_concordance,
        [[], [], params.bins, [], []],
        params.min_val_gl, params.min_val_dp
    )
    ch_versions = ch_versions.mix(GLIMPSE2_CONCORDANCE.out.versions.first())

    ch_multiqc_files = ch_multiqc_files.mix(GLIMPSE2_CONCORDANCE.out.errors_cal.map{meta, txt -> [txt]})
    ch_multiqc_files = ch_multiqc_files.mix(GLIMPSE2_CONCORDANCE.out.errors_grp.map{meta, txt -> [txt]})
    ch_multiqc_files = ch_multiqc_files.mix(GLIMPSE2_CONCORDANCE.out.errors_spl.map{meta, txt -> [txt]})
    ch_multiqc_files = ch_multiqc_files.mix(GLIMPSE2_CONCORDANCE.out.rsquare_grp.map{meta, txt -> [txt]})
    ch_multiqc_files = ch_multiqc_files.mix(GLIMPSE2_CONCORDANCE.out.rsquare_spl.map{meta, txt -> [txt]})
    ch_multiqc_files = ch_multiqc_files.mix(GLIMPSE2_CONCORDANCE.out.rsquare_per_site.map{meta, txt -> [txt]})

    GUNZIP(GLIMPSE2_CONCORDANCE.out.errors_grp)
    ch_versions = ch_versions.mix(GUNZIP.out.versions.first())

    ADD_COLUMNS(GUNZIP.out.gunzip)
    ch_versions = ch_versions.mix(ADD_COLUMNS.out.versions.first())

    GAWK(
        ADD_COLUMNS.out.txt
            .map{meta, txt -> [["id":"TestQuality"], txt]}
            .groupTuple(),
        []
    )
    ch_versions = ch_versions.mix(GAWK.out.versions.first())

    emit:
    stats           = GAWK.out.output             // [ [all], txt ]
    versions        = ch_versions                 // channel: [ versions.yml ]
    multiqc_files   = ch_multiqc_files
}
