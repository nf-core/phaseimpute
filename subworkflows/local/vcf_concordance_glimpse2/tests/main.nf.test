nextflow_workflow {

    name "Test Subworkflow VCF_CONCORDANCE_GLIMPSE2"
    script "../main.nf"
    config "./nextflow.config"

    workflow "VCF_CONCORDANCE_GLIMPSE2"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_concordance_glimpse2"
    tag "vcf_concordance_glimpse2"

    tag "bcftools"
    tag "bcftools/index"
    tag "glimpse"
    tag "glimpse/phase"
    tag "glimpse/concordance"

    test("vcf_concordance_glimpse2") {
        when {
            workflow {
                """
                allele_freq = Channel.fromList([
                    [
                        [panel:'1000GP', chr:'21'], // meta map
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.sites.vcf.gz",checkIfExists:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.sites.vcf.gz.csi",checkIfExists:true)
                    ]
                ])
                truth = Channel.fromList([
                    [[id:'NA12878'], // meta map
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bcf",checkIfExists:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bcf.csi",checkIfExists:true)],
                    [[id:'NA19401'], // meta map
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bcf",checkIfExists:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bcf.csi",checkIfExists:true)],
                ])
                estimate = Channel.fromList([
                    [[id:'NA12878'], // meta map
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s_imputed.bcf",checkIfExists:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s_imputed.bcf.csi",checkIfExists:true)],
                    [[id:'NA19401'], // meta map
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s_imputed.bcf",checkIfExists:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s_imputed.bcf.csi",checkIfExists:true)]
                ])
                region = Channel.of([ [chr:'21', region:'chr21:16570000-16610000'],  "chr21:16570000-16610000"])
                input[0] = estimate
                input[1] = truth
                input[2] = allele_freq
                input[3] = region
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

    }
}
