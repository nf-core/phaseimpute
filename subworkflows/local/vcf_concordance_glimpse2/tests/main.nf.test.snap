{
    "vcf_concordance_glimpse2": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "TestQuality"
                        },
                        "TestQuality.txt:md5,0e199fabdd479db15849394e148bd1e1"
                    ]
                ],
                "1": [
                    "versions.yml:md5,11b3f37c4921334a4e4881b5fb466802",
                    "versions.yml:md5,185cbca93dea56aa1c36636716a6ac89",
                    "versions.yml:md5,8bc5d870dd6bf1b13e84535e0b71ce39",
                    "versions.yml:md5,cb85a5c4bc22951d41b21bc660e89637"
                ],
                "2": [
                    [
                        "NA12878.error.cal.txt.gz:md5,33c10b18d109bc6fd53fc2594e18a844"
                    ],
                    [
                        "NA12878.error.grp.txt.gz:md5,36f3fc826367dc6e89e95b58d90e94eb"
                    ],
                    [
                        "NA12878.error.spl.txt.gz:md5,e7b0730fba0dc331de80ed911fcb6dcf"
                    ],
                    [
                        "NA12878.rsquare.grp.txt.gz:md5,4e2d341610d908d3f78685388f7d41b8"
                    ],
                    [
                        "NA12878.rsquare.spl.txt.gz:md5,b9c53f1ad2c5fc5243ccbddb4a8b4260"
                    ],
                    [
                        "NA19401.error.cal.txt.gz:md5,46a6d42d469328096a66b8b172bce6d7"
                    ],
                    [
                        "NA19401.error.grp.txt.gz:md5,0a37ae2f1a5ce339a798f0438d80fd82"
                    ],
                    [
                        "NA19401.error.spl.txt.gz:md5,efd6bd3bf0c5530f0cf9b5ad055abe49"
                    ],
                    [
                        "NA19401.rsquare.grp.txt.gz:md5,8161d852ad5a77d5b83236960497cfe1"
                    ],
                    [
                        "NA19401.rsquare.spl.txt.gz:md5,fb9ff57181626f06a1947b62ad18f97d"
                    ]
                ],
                "multiqc_files": [
                    [
                        "NA12878.error.cal.txt.gz:md5,33c10b18d109bc6fd53fc2594e18a844"
                    ],
                    [
                        "NA12878.error.grp.txt.gz:md5,36f3fc826367dc6e89e95b58d90e94eb"
                    ],
                    [
                        "NA12878.error.spl.txt.gz:md5,e7b0730fba0dc331de80ed911fcb6dcf"
                    ],
                    [
                        "NA12878.rsquare.grp.txt.gz:md5,4e2d341610d908d3f78685388f7d41b8"
                    ],
                    [
                        "NA12878.rsquare.spl.txt.gz:md5,b9c53f1ad2c5fc5243ccbddb4a8b4260"
                    ],
                    [
                        "NA19401.error.cal.txt.gz:md5,46a6d42d469328096a66b8b172bce6d7"
                    ],
                    [
                        "NA19401.error.grp.txt.gz:md5,0a37ae2f1a5ce339a798f0438d80fd82"
                    ],
                    [
                        "NA19401.error.spl.txt.gz:md5,efd6bd3bf0c5530f0cf9b5ad055abe49"
                    ],
                    [
                        "NA19401.rsquare.grp.txt.gz:md5,8161d852ad5a77d5b83236960497cfe1"
                    ],
                    [
                        "NA19401.rsquare.spl.txt.gz:md5,fb9ff57181626f06a1947b62ad18f97d"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "TestQuality"
                        },
                        "TestQuality.txt:md5,0e199fabdd479db15849394e148bd1e1"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,11b3f37c4921334a4e4881b5fb466802",
                    "versions.yml:md5,185cbca93dea56aa1c36636716a6ac89",
                    "versions.yml:md5,8bc5d870dd6bf1b13e84535e0b71ce39",
                    "versions.yml:md5,cb85a5c4bc22951d41b21bc660e89637"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T12:39:49.843552307"
    }
}